allennlp/models/reading_comprehension/document_qa.py

```diff
@@ -0,0 +1,401 @@
+import logging
+from typing import Any, Dict, List
+
+import torch
+import torch.nn.functional as F
+from torch.autograd import Variable
+
+from allennlp.common import Params, squad_eval
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import Seq2SeqEncoder, TimeDistributed, TextFieldEmbedder, TriLinearAttention
+from allennlp.modules.time_distributed import squash, unsquash
+from allennlp.nn import InitializerApplicator, util
+from allennlp.nn.loss import SharedNormLoss
+from allennlp.training.metrics import Average, BooleanAccuracy, CategoricalAccuracy
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+def repeat_question(question_tensor: Variable, num_paragraphs: int) -> Variable:
+    """
+    Turns a (batch_size, num_tokens, input_dim) tensor representing a question into a
+    (batch_size * num_paragraphs, num_tokens, input_dim) tensor to be paired with each paragraph.
+    """
+    if num_paragraphs == 1:
+        return question_tensor
+    old_size = question_tensor.size()
+    # question_tensor is (batch_size, sequence_length, input_dim)
+    # .unsqueeze(1) is (batch_size, 1, sequence_length, input_dim)
+    # .repeat() is (batch_size, num_paragraphs, sequence_length, input_dim)
+    # .view() is (batch_size * num_paragraphs, sequence_length, input_dim)
+    tensor = question_tensor.unsqueeze(1).repeat(torch.Size([1] + [num_paragraphs] + [1] * (len(old_size) - 1)))
+    tensor = tensor.view(torch.Size([old_size[0] * num_paragraphs]) + old_size[1:])
+    return tensor
+
+@Model.register("document-qa")
+class DocumentQa(Model):
+    """
+    This class implements Christopher Clark and Matt Gardner's
+    `Multi-Paragraph Reading Comprehension
+    <https://www.semanticscholar.org/paper/Simple-and-Effective-Multi-Paragraph-Reading-Compr-Clark-Gardner/10201edcf04a102a1c4f8ed7107562a13148dd81>`_
+    for answering reading comprehension questions.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 phrase_layer: Seq2SeqEncoder,
+                 residual_encoder: Seq2SeqEncoder,
+                 span_start_encoder: Seq2SeqEncoder,
+                 span_end_encoder: Seq2SeqEncoder,
+                 initializer: InitializerApplicator,
+                 dropout: float = 0.2,
+                 mask_lstms: bool = True) -> None:
```

Comment:
What is the scenario where you would not want to mask the output of the lstm?

Reply:
In some early experiments I did with BiDAF, I found that omitting the masking in the LSTMs sped things up a whole lot (as it lets us skip some expensive sorting before passing things off to CUDA), and it didn't really affect accuracy.  So this flag would let you trade off a little bit of correctness for a whole bunch of speed.  @nelson-liu recently tried to reproduce the speedup I saw, though, and didn't see it.  We're not sure why - maybe it doesn't make as much difference with newer hardware, or maybe some other code change we did in the interim made it so we're always sorting anyway, or something.